Improve the SciJava Ops benchmarks

ctrueden opened the pull request:
* Make the underlying op implementation as trivial as possible, while still preventing the JIT from optimizing away method calls completely.
* Merge all benchmarking classes into a single `Benchmarks.java`, to avoid repeatedly benchmarking the same thing across the classes.
* Add static/non-Ops benchmarks that also perform adaptation, conversion, and both, so that we can compare apples to apples there.
* Update the Python script accordingly to consume the revised benchmarking output, and generate the same charts as before, with a couple of additional bars where appropriate.
* Switch the unit of measure from milliseconds to microseconds.
* Fix a couple of trivial issues with the Benchmarks doc page.

ctrueden committed:
Simplify byte inversion code

ctrueden committed:
Use more succinct fast-failing service accessor

ctrueden committed:
scijava-persist: tidy the POM

ctrueden committed:
Tweak SciJava plugin metadata aggregation hack

Generally speaking, it is not correct to mix the contents of
target/classes/META-INF/org.scijava.plugin.Plugin into
target/test-classes/META-INF/org.scijava.plugin.Plugin, because it
creates duplicate plugin entries across the two classpath elements.

The reason for the script-maven-plugin hack is to avoid module-path
woes, where (if I understand correctly) the target/classes
directory is on the module path but the target/test-classes
directory is on the classpath, which somehow results in the unit
tests not being able to discover plugins in the target/classes
folder, even though the src/main/java/module-info.java includes
e.g. `opens org.scijava.persist to org.scijava`. And similarly for
SciJava Ops, not being able to notice the `target/classes/ops.yaml`
from unit tests. This is an unfortunate situation, which we have
tried to resolve by introducing a `src/test/java/module-info.java`
as well, but have never quite been able to make it work correctly.

Strangely, the scijava-ops-benchmark component introduces a single
`@Plugin` of type `net.imagej.ops.Op` in its main scope, which as
per the script hack gets propagated into both target/classes and
target/test-classes... but then the unit tests *do* see both of
these files and read them both, resulting in an ImageJ Ops
environment containing that Op plugin twice, which results in the
dreaded "Multiple ops of priority 0.0" error.

I do not understand why the behavior differs between modules.
Nonetheless, this commit seeks to avoid the issue by only combining
META-INF/org.scijava.plugin.Plugin metadata files between
target/classes and target/test-classes for the one component that
actually needs it, which is scijava-persist.

ctrueden committed:
Rework and simplify the benchmarks

* Make the op implementation as trivial as possible, while still
  preventing the JIT from optimizing away method calls completely.

* Merge all benchmarking classes into a single Benchmarks.java,
  to avoid repeatedly benchmarking the same thing across the classes.

* Add static/non-Ops benchmarks that also perform adaptation,
  conversion, and both, so that we can compare apples to apples there.

* Update the Python script accordingly to consume the revised
  benchmarking output, and generate the same charts as before,
  with a couple of additional bars where appropriate.

* Fix a couple of trivial issues with the Benchmarks doc page.

ctrueden committed:
Purge SCIFIO dependency from scijava-ops-opencv

It introduces an org.scijava.Context, which is not necessary for the
test in question. The previous test also hit the network on every run.

ctrueden commented:
The question of why cached matches are slower than non-cached ones is still open, and I plan to investigate soon. But that should not hold up the merger of this PR, which I feel is a step forward for the benchmarks regardless.

gselzer commented on docs/ops/doc/index.rst:
```diff
@@ -22,6 +22,7 @@ The combination of these libraries allows declarative image analysis workflows,
 
    CallingOps
    SearchingForOps
+   ScriptingInFiji
```
This file is now duplicated in the list

gselzer commented on docs/ops/bin/benchmark.sh:
```diff
@@ -65,4 +65,4 @@ fi
 echo
 echo '=== GRAPHING THE RESULTS ==='
-"$pythonPath" graph_results.py
+#"$pythonPath" graph_results.py
```
Should this be uncommented?

gselzer commented on scijava-ops-benchmarks/src/main/java/org/scijava/ops/benchmarks/Benchmarks.java:
```diff
@@ -0,0 +1,189 @@
+	 * @param state the state between benchmarks
+	 */
+	@Benchmark
+	public void sjOpsNoCache(final BenchmarkState state) {
```
It's a little strange that this benchmark and the one below all ignore the cache, but only this one has `NoCache` in the name. But it's probably fine.

ctrueden replied:
I know, I don't like it. I was going to rename it to be more consistent, but forgot. I'll take care of it!

gselzer reviewed (commented):
Super awesome! Couple small suggestions

gselzer commented:
@ctrueden it would also be nice to have a little section at the end of the RTD benchmark page with the combined graph - otherwise, the JSON you've included will be inaccessible.

hinerm reviewed (commented):
Looks good to me with Gabe's comments addressed and the caching investigation.

ctrueden commented:
Thanks! I've addressed all the feedback points:

* No more duplicate `ScriptingInFiji`
* No more commented-out python script execution step
* More consistent naming of the sjOps test methods

About adding the fifth "combined metrics" chart: it was already there, so I think all is well? Because I had slipped it onto the main branch a few days ago, before this more major reworking. :sweat_smile: 

Still working on updating the `Benchmarks.rst` to be correct, and additional adaptation tests, and caching investigation.

gselzer committed:
OpBuilder: pass hints for inplaces

hinerm commented:
@gselzer I removed inplaces and made the test Op a `Computer` that's adapted to a `Function`.

However, the nullary computer does not match. The `testSciJavaOps` and `testConversion` tests fail:

```
org.scijava.ops.api.OpMatchingException: No match found! Perhaps you meant: 
benchmark.increment:
	- org.scijava.ops.benchmarks.OpsToBenchmark.incrementByte(byte[])
		 Increments a byte value.
		
	
		> mutable1 : @MUTABLE byte[]
			array containing the byte to increment
...

Request:
-	Name: "benchmark.increment", Types: org.scijava.function.Computers.Arity0<[B>
Input Types: 
		* byte[]
Output Type: 
		* byte[]
```

gselzer committed:
Update Benchmarks.rst
